standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/columnstats/DecimalComparatorTest.java

```diff
@@ -0,0 +1,446 @@
+    check.accept(10, -512);
+    check.accept(10, -1023);
+    check.accept(11, -1024);
+
+    // log2(5800) is about 12.5, bitLog floor(log(...))+1
+    check.accept(13, 5800);
+    check.accept(13, -5800);
+  }
+
+  @Test
+  public void testBitLog2() {
+    interface Helper {
+      void accept(int expected, String input, byte pad);
+    }
+    Helper check = (expected, input, pad) -> assertEquals(expected, bitLog(FORMAT.parseHex(input), pad));
+
+    check.accept(8, "FF 16", PAD_NEG);
+
+    check.accept(0, "00 00 00", PAD_POS);
+    check.accept(1, "00 00 01", PAD_POS);
+    check.accept(2, "00 00 02", PAD_POS);
+    check.accept(3, "00 00 04", PAD_POS);
+    check.accept(4, "00 00 08", PAD_POS);
+    check.accept(5, "00 00 10", PAD_POS);
+    check.accept(5, "00 00 11", PAD_POS);
+    check.accept(9, "00 01 11", PAD_POS);
+    check.accept(17, "01 11 11", PAD_POS);
+    check.accept(17, "01 00 00", PAD_POS);
+    check.accept(17, "00 00 00 01 00 00", PAD_POS);
+
+    check.accept(1, "FF FF FF", PAD_NEG);
+    check.accept(2, "FF FF FE", PAD_NEG);
+    check.accept(2, "FF FF FD", PAD_NEG);
+    check.accept(3, "FF FF FB", PAD_NEG);
+    check.accept(4, "FF FF F7", PAD_NEG);
+    check.accept(5, "FF FF EF", PAD_NEG);
+    check.accept(5, "FF FF EE", PAD_NEG);
+    check.accept(9, "FF FE EE", PAD_NEG);
+    check.accept(17, "FE EE EE", PAD_NEG);
+    check.accept(17, "FE FF FF", PAD_NEG);
+    check.accept(17, "FF FF FF FE FF FF", PAD_NEG);
+  }
+
+  @Test
+  public void testBitLenPostcondition() {
+    // define helper which executes the asserts
+    Consumer<BigInteger> check = x -> {
+      int bl = bitLog(x.toByteArray(), x.signum() == -1 ? PAD_NEG : PAD_POS);
+      int log = BigIntegerMath.log2(x.abs(), RoundingMode.DOWN);
+      if (log < 0) {
+        fail("Log may not be negative: " + x);
+      }
+
+      if (log + 1 != bl) {
+        fail(x.toString());
+      }
+      // check inequalities
+      if (!(bl - 1 <= log)) {
+        fail(x.toString());
+      }
+      if (!(log < bl)) {
+        fail(x.toString());
+      }
+    };
+
+    // check all integers from -18 to 18 (inclusive)
+    for (int i = -18; i <= 18; i++) {
+      if (i == 0) {
+        continue;
+      }
+      BigInteger x = BigInteger.valueOf(i);
+      check.accept(x);
+    }
+
+    // check powers of two from 2^5 to 2^200 (about 60 decimal digits)
+    for (int i = 5; i < 200; i++) {
+      BigInteger p = BigInteger.TWO.pow(i);
+      // check negative and positive powers of two: -2^p and +2^p
+      for (int neg = 0; neg < 2; neg++) {
+        if (neg == 1) {
+          p = p.negate();
+        }
+
+        // check the neighborhood of each power of two
+        for (int j = -2; j <= 2; j++) {
+          BigInteger x = p.add(BigInteger.valueOf(j));
+          if (x.compareTo(BigInteger.ZERO) == 0) {
+            continue;
+          }
+          check.accept(x);
+        }
+      }
+    }
+  }
+
+  @Test
+  public void testRandomized() {
```

Comment:
Randomized tests could lead to flakyness and hard to reproduce.

Reply:
The test includes the seed in the error message. I've executed the test over 25000 times, and it passed without a failure. A randomized test has the advantage to cover more numbers. If it fails, then the test would be marked as flaky, but a discrepancy between BigDecimal.compareTo and the comparator of the PR has been found, so it should be investigated. I think that's a reasonable price to pay for finding a (rare) bug.